tls/extensions/s2n_client_alpn.c

```diff
@@ -71,35 +68,41 @@ static int s2n_client_alpn_recv(struct s2n_connection *conn, struct s2n_stuffer
         return S2N_SUCCESS;
     }
 
+    uint16_t size_of_all;
     POSIX_GUARD(s2n_stuffer_read_uint16(extension, &size_of_all));
     if (size_of_all > s2n_stuffer_data_available(extension) || size_of_all < 3) {
         /* Malformed length, ignore the extension */
         return S2N_SUCCESS;
     }
 
-    struct s2n_blob client_app_protocols = { 0 };
-    client_app_protocols.size = size_of_all;
-    client_app_protocols.data = s2n_stuffer_raw_read(extension, size_of_all);
-    POSIX_ENSURE_REF(client_app_protocols.data);
-
-    /* Find a matching protocol */
-    POSIX_GUARD(s2n_stuffer_init(&server_protos, server_app_protocols));
-    POSIX_GUARD(s2n_stuffer_skip_write(&server_protos, server_app_protocols->size));
-
-    while (s2n_stuffer_data_available(&server_protos) > 0) {
-        struct s2n_blob server_protocol = { 0 };
-        POSIX_ENSURE(s2n_result_is_ok(s2n_protocol_preferences_read(&server_protos, &server_protocol)),
-                S2N_ERR_BAD_MESSAGE);
-
-        bool is_match = false;
-        POSIX_ENSURE(s2n_result_is_ok(s2n_protocol_preferences_contain(&client_app_protocols, &server_protocol, &is_match)),
-                S2N_ERR_BAD_MESSAGE);
+    POSIX_GUARD(s2n_select_supported_protocol(conn, server_app_protocols, extension));
+    return S2N_SUCCESS;
+}
 
-        if (is_match) {
-            POSIX_CHECKED_MEMCPY(conn->application_protocol, server_protocol.data, server_protocol.size);
-            conn->application_protocol[server_protocol.size] = '\0';
+int s2n_select_supported_protocol(struct s2n_connection *conn, struct s2n_blob *protocol_preferences, struct s2n_stuffer *extension)
+{
+    uint16_t wire_size = s2n_stuffer_data_available(extension);
+    struct s2n_blob peer_protocols = { 0 };
+    peer_protocols.size = wire_size;
+    peer_protocols.data = s2n_stuffer_raw_read(extension, wire_size);
+    POSIX_ENSURE_REF(peer_protocols.data);
+
+    struct s2n_stuffer protocols = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&protocols, protocol_preferences));
+    POSIX_GUARD(s2n_stuffer_skip_write(&protocols, protocol_preferences->size));
+
+    while(s2n_stuffer_data_available(&protocols) > 0) {
+        struct s2n_blob protocol = { 0 };
+        POSIX_ENSURE(s2n_result_is_ok(s2n_protocol_preferences_read(&protocols, &protocol)), S2N_ERR_BAD_MESSAGE);
+        
+        bool match_found = false;
+        POSIX_ENSURE(s2n_result_is_ok(s2n_protocol_preferences_contain(&peer_protocols, &protocol, &match_found)), S2N_ERR_BAD_MESSAGE);
```

Comment:
Are we doing POSIX_ENSURE instead of POSIX_GUARD_RESULT because we really want to return S2N_ERR_BAD_MESSAGE?

Maybe the extensions framework should translate any S2N_ERR_SAFETY to S2N_ERR_BAD_MESSAGE. Also not a blocker for your PR.